topology/probes/k8s/cache.go

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package k8s
+
+import (
+	"time"
+
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology/graph"
+
+	api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ResourceHandler is used to map Kubernetes resources to objets in the graph
+type ResourceHandler interface {
+	Map(obj interface{}) (graph.Identifier, graph.Metadata, error)
+	Dump(obj interface{}) string
+	IsTopLevel() bool
+}
+
+type DefaultResourceHandler struct {
+}
+
+func (h *DefaultResourceHandler) IsTopLevel() bool {
+	return false
+}
+
+// ResourceCache describes a cache for a specific kind of Kubernetes resource.
+// It is in charge of listening Kubernetes events and creating the
+// according resource in the graph with the informations returned by
+// the associated resource handler
+type ResourceCache struct {
+	*graph.GraphEventHandler
+	cache          cache.Indexer
+	controller     cache.Controller
+	stopController chan (struct{})
+	handler        ResourceHandler
+	graph          *graph.Graph
+}
+
+func (c *ResourceCache) list() []interface{} {
+	return c.cache.List()
+}
+
+func (c *ResourceCache) getByKey(namespace, name string) interface{} {
+	key := ""
+	if len(namespace) > 0 {
+		key = namespace + "/"
+	}
+	key += name
+	if obj, found, _ := c.cache.GetByKey(key); found {
+		return obj
+	}
+	return nil
+}
+
+func (c *ResourceCache) getByNode(node *graph.Node) interface{} {
+	namespace, _ := node.GetFieldString("Namespace")
+	name, _ := node.GetFieldString("Name")
+	if name == "" {
+		return nil
+	}
+	return c.getByKey(namespace, name)
+}
+
+func (c *ResourceCache) getByNamespace(namespace string) []interface{} {
+	if namespace == api.NamespaceAll {
+		return c.list()
+	}
+
+	objects, _ := c.cache.ByIndex("namespace", namespace)
+	return objects
+}
+
+func (c *ResourceCache) getBySelector(g *graph.Graph, namespace string, selector *metav1.LabelSelector) []metav1.Object {
```

Comment:
do we want to support the case in which `selector == nil` (and then just return all items in namespace)